In [8]:
import numpy as np
import torch
import torch.nn as nn
import torch.utils.checkpoint as checkpoint
from PIL import Image 
import torchvision.transforms as transforms 
import matplotlib.pyplot as plt
import seaborn as sns
import os
from torch.utils.data import Dataset, DataLoader
import torch.nn.functional as F

Prepare data

In [9]:

def get_sub_path(path):
    sub_path = []
    if isinstance(path, list):
        for p in path:
            if os.path.isdir(p):
                for file in os.listdir(p):
                    sub_path.append(os.path.join(p, file))
            else:
                continue
    else:
        for file in os.listdir(path):
            sub_path.append(os.path.join(path, file))
    return sub_path

def divide_list(list, n):
    for i in range(0, len(list), n):
        yield list[i:i + n]
        
def std(input):
    if input.max() == 0:
        return input
    else:
        result = (input-input.min()) / (input.max()-input.min())
        return result

In [10]:
feature_list = ['IR_drop_features_decompressed/power_i', 'IR_drop_features_decompressed/power_s', 
        'IR_drop_features_decompressed/power_sca', 'IR_drop_features_decompressed/power_all']
label_list = ['IR_drop_features_decompressed/IR_drop']

datapath = './CircuitNet-N28/'
# datapath = '../../CircuitNet/CircuitNet-N28/'
name_list = get_sub_path(os.path.join(datapath, feature_list[-1]))
n_list = divide_list(name_list, 1000)

In [11]:
class PowerDataset(Dataset):
    def __init__(self, root_dir, target_size=(224, 224)):
        self.root_dir = root_dir
        self.feature_dirs = ['power_i', 'power_s', 'power_sca', 'Power_all']
        self.label_dir = 'IR_drop'
        self.target_size = target_size
        # Collect all the feature and label file paths
        self.data = []
        i=0
        for case_name in os.listdir(os.path.join(root_dir, self.feature_dirs[0])):
            feature_paths = [os.path.join(root_dir, feature_dir, case_name) for feature_dir in self.feature_dirs]
            label_path = os.path.join(root_dir, self.label_dir, case_name)
            if all(os.path.exists(fp) for fp in feature_paths) and os.path.exists(label_path):
                self.data.append((feature_paths, label_path))
            i+=1
            # we intentionally limit the size of our dataset to 100
            if i>99:
                break

    def __len__(self):
        return len(self.data)
    
    def __getitem__(self, idx):
        feature_paths, label_path = self.data[idx]
        features = []
        
        for fp in feature_paths:
            feature = np.load(fp)
            feature = torch.tensor(feature, dtype=torch.float32)
            feature = F.interpolate(feature.unsqueeze(0).unsqueeze(0), size=self.target_size, mode='nearest').squeeze(0).squeeze(0)
            feature = std(feature)
            features.append(feature)
            
        features = torch.stack(features, dim=0)
        
        # Load and process label file
        label = np.load(label_path)
        label = torch.tensor(label, dtype=torch.float32)
        label = F.interpolate(label.unsqueeze(0).unsqueeze(0), size=self.target_size, mode='nearest').squeeze(0).squeeze(0)
        label = label.clamp(1e-6, 50)
        label = (torch.log10(label)+6)/ (np.log10(50)+6)
        
        return features, label
    
root_dir = './CircuitNet-N28/IR_drop_features_decompressed/'
dataset = PowerDataset(root_dir)

Size of datapoint

In [19]:
feature_dirs = ['power_i', 'power_s', 'power_sca', 'Power_all']
label_dir = 'IR_drop'
c=0
r1 = 0 # number of datapoints in [220, 280]
r2 = 0 # number of datapoints in [200, 300]
r3 = 0 # number of datapoints in [200, 260]
for case_name in os.listdir(os.path.join(root_dir, feature_dirs[0])):
    feature_paths = [os.path.join(root_dir, feature_dir, case_name) for feature_dir in feature_dirs]
    label_path = os.path.join(root_dir, label_dir, case_name)
    c+=1
    # print(feature_paths)
    # in one data point, all power_? has same shape
    x = np.load(feature_paths[0])
    x = torch.tensor(x)
    if x.shape[0] >= 220 and x.shape[0] <= 280 and x.shape[1] >= 220 and x.shape[1] <= 280:
        r1+=1
    if x.shape[0] >= 200 and x.shape[0] <= 300 and x.shape[1] >= 200 and x.shape[1] <= 300:
        r2+=1
    if x.shape[0] >= 220 and x.shape[0] <= 290 and x.shape[1] >= 220 and x.shape[1] <= 290:
        r3+=1
    # print(x.shape)
    # for i in range (len(feature_paths)):
    #     x = np.load(feature_paths[i])
    #     x = torch.tensor(x)
    #     print(x.shape)
print("{:.2f}% of data points fall into 220*220 ~ 280*280".format(r1/c * 100) )
print("{:.2f}% of data points fall into 220*220 ~ 290*290".format(r3/c * 100) )
print("{:.2f}% of data points fall into 200*200 ~ 300*300".format(r2/c * 100) )
print(c)

74.23% of data points fall into 220*220 ~ 280*280
87.07% of data points fall into 220*220 ~ 290*290
95.66% of data points fall into 200*200 ~ 300*300
10242


Value range of datapoint

In [20]:
def countlarger(tensor):
    return torch.sum(tensor > 1).item()
def countbetween(tensor):
    return torch.sum((tensor > 0) & (tensor < 1)).item()
def count_ones(tensor):
    return torch.sum(tensor == 1).item()
def count_zeros(tensor):
    return torch.sum(tensor == 0).item()

def print_non_zeros(tensor):
    non_zeros = tensor[tensor != 0]
    print("Non-zero elements in the tensor:")
    print(non_zeros)

其实只有10242个file。标签最大是10370，中间有些丢失了。


                  value = 0, 0 < value < 1, value = 1, value > 1


power_i:          502616514 222652424 0 0


power_s:          502754241 222514697 0 0


power_sca:        485297920 239971018 0 0


power_all:        485297920 239971018 0 0

In [21]:
feature_dirs = ['power_i', 'power_s', 'power_sca', 'Power_all']
label_dir = 'IR_drop'
c=0
r1 = 0 # value = 0
r2 = 0 # 0 < value < 1
r3 = 0 # value = 1
r4 = 0 # value > 1
for case_name in os.listdir(os.path.join(root_dir, feature_dirs[0])):
    # if case_name.startswith('1'):
    #     print(case_name)
    feature_paths = [os.path.join(root_dir, feature_dir, case_name) for feature_dir in feature_dirs]
    label_path = os.path.join(root_dir, label_dir, case_name)
    c+=1
    # print(feature_paths)
    # =================================================================================================
    # in one data point, all power_? has same shape
    x = np.load(feature_paths[1])
    x = torch.tensor(x)
    # print(x.shape)
    x1 = count_zeros(x)
    x2 = countbetween(x)
    x3 = count_ones(x)
    x4 = countlarger(x)
    r1 += x1
    r2 += x2
    r3 += x3
    r4 += x4
    # print(f"> 0: {x1}, 0 < x < 1: {x2}, > 1: {x3}")
    # if (c == 10): break
    # for i in range (len(feature_paths)):
    #     x = np.load(feature_paths[i])
    #     x = torch.tensor(x)
    #     print(x.shape)

print(r1, r2, r3, r4)
print(c)

502754241 222514697 0 0
10242


In [12]:
dataloader = DataLoader(dataset, batch_size=4, shuffle=True)
# number of datapoints / batch_size = len(dataloader)
print(len(dataloader))        # 25
print(len(dataset))           # 100

c=0
for features, labels in dataloader:
    print(features.shape, labels.shape)
    if (c == 10): break
    c+=1

25
100
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])
torch.Size([4, 4, 224, 224]) torch.Size([4, 224, 224])


In [23]:
from swintransformer import *

model_name = 'swin_base_patch4_window7_224'
model = init_model(model_name, input_channels=4, num_classes=0, pretrained=True)

In [24]:

model.eval()
with torch.no_grad():
    ir_prediction = model(features)

In [25]:
ir_prediction.shape

torch.Size([4, 1, 224, 224])

In [26]:
import utils.losses as losses
import torch.optim as optim
from tqdm import tqdm
from math import cos, pi
model.train()

def checkpoint(model, epoch, save_path):
    if not os.path.exists(save_path):
        os.makedirs(save_path)
    model_out_path = f"./{save_path}/swinTransformer_iters_{epoch}.pth"
    torch.save({'state_dict': model.state_dict()}, model_out_path)
    print("Checkpoint saved to {}".format(model_out_path))

class CosineRestartLr(object):
    def __init__(self,
                 base_lr,
                 periods,
                 restart_weights = [1],
                 min_lr = None,
                 min_lr_ratio = None):
        self.periods = periods
        self.min_lr = min_lr
        self.min_lr_ratio = min_lr_ratio
        self.restart_weights = restart_weights
        super().__init__()

        self.cumulative_periods = [
            sum(self.periods[0:i + 1]) for i in range(0, len(self.periods))
        ]

        self.base_lr = base_lr

    def annealing_cos(self, start: float,
                    end: float,
                    factor: float,
                    weight: float = 1.) -> float:
        cos_out = cos(pi * factor) + 1
        return end + 0.5 * weight * (start - end) * cos_out

    def get_position_from_periods(self, iteration: int, cumulative_periods):
        for i, period in enumerate(cumulative_periods):
            if iteration < period:
                return i
        raise ValueError(f'Current iteration {iteration} exceeds '
                        f'cumulative_periods {cumulative_periods}')


    def get_lr(self, iter_num, base_lr: float):
        target_lr = self.min_lr  # type:ignore

        idx = self.get_position_from_periods(iter_num, self.cumulative_periods)
        current_weight = self.restart_weights[idx]
        nearest_restart = 0 if idx == 0 else self.cumulative_periods[idx - 1]
        current_periods = self.periods[idx]

        alpha = min((iter_num - nearest_restart) / current_periods, 1)
        return self.annealing_cos(base_lr, target_lr, alpha, current_weight)

    
    def _set_lr(self, optimizer, lr_groups):
        if isinstance(optimizer, dict):
            for k, optim in optimizer.items():
                for param_group, lr in zip(optim.param_groups, lr_groups[k]):
                    param_group['lr'] = lr
        else:
            for param_group, lr in zip(optimizer.param_groups,
                                        lr_groups):
                param_group['lr'] = lr

    def get_regular_lr(self, iter_num):
        return [self.get_lr(iter_num, _base_lr) for _base_lr in self.base_lr]  # iters

    def set_init_lr(self, optimizer):
        for group in optimizer.param_groups:  # type: ignore
            group.setdefault('initial_lr', group['lr'])
            self.base_lr = [group['initial_lr'] for group in optimizer.param_groups  # type: ignore
        ]


# Build loss
loss = losses.__dict__['L1Loss']()

# arg_dict = {'task': 'irdrop_mavi', 'save_path': 'work_dir/irdrop_mavi/', 'pretrained': None, 'max_iters': 2500, 'plot_roc': False, 'arg_file': None, 'cpu': True, 'dataroot': 'CircuitNet-N28/training_set/IR_drop', 'ann_file_train': './files/train_N28.csv', 'ann_file_test': './files/test_N28.csv', 'dataset_type': 'IRDropDataset', 'batch_size': 2, 'model_type': 'MAVI', 'in_channels': 1, 'out_channels': 4, 'lr': 0.0002, 'weight_decay': 0.01, 'loss_type': 'L1Loss', 'eval_metric': ['NRMS', 'SSIM'], 'threshold': 0.9885, 'ann_file': './files/train_N28.csv', 'test_mode': False}
arg_dict = {'task': 'irdrop_mavi', 'save_path': 'work_dir/irdrop_mavi/', 'pretrained': None, 'max_iters': 5000, 'plot_roc': True, 'arg_file': None, 'cpu': True, 'dataroot': 'CircuitNet-N28/training_set/IR_drop', 'ann_file_train': './files/train_N28.csv', 'ann_file_test': './files/test_N28.csv', 'dataset_type': 'IRDropDataset', 'batch_size': 2, 'model_type': 'MAVI', 'in_channels': 1, 'out_channels': 4, 'lr': 0.00005, 'weight_decay': 0.01, 'loss_type': 'L1Loss', 'eval_metric': ['NRMS', 'SSIM'], 'threshold': 0.9885, 'ann_file': './files/train_N28.csv', 'test_mode': False}

# Build Optimzer
optimizer = optim.AdamW(model.parameters(), lr=arg_dict['lr'],  betas=(0.9, 0.999), weight_decay=arg_dict['weight_decay'])

# Build lr scheduler
cosine_lr = CosineRestartLr(arg_dict['lr'], [arg_dict['max_iters']], [1], 1e-7)
cosine_lr.set_init_lr(optimizer)

epoch_loss = 0
iter_num = 0
print_freq = 100
# save_freq = 10000
save_freq = 100

while iter_num < arg_dict['max_iters']:
    with tqdm(total=print_freq) as bar:
        # for feature, label, _ in dataset:     
        for feature, label in dataloader:   
            if arg_dict['cpu']:
                input, target = feature, label
            else:
                input, target = feature.cuda(), label.cuda()

            regular_lr = cosine_lr.get_regular_lr(iter_num)
            cosine_lr._set_lr(optimizer, regular_lr)

            prediction = model(input)
            # print(input.shape)

            optimizer.zero_grad()
            pixel_loss = loss(prediction, target)

            epoch_loss += pixel_loss.item()
            pixel_loss.backward()
            optimizer.step()

            iter_num += 1
            
            bar.update(1)

            if iter_num % print_freq == 0:
                break

    print("===> Iters[{}]({}/{}): Loss: {:.4f}".format(iter_num, iter_num, arg_dict['max_iters'], epoch_loss / print_freq))
    if iter_num % save_freq == 0:
        checkpoint(model, iter_num, arg_dict['save_path'])
    epoch_loss = 0

  0%|          | 0/100 [00:00<?, ?it/s]/Users/ziyichu/Desktop/Research/CircuitNet/routability_ir_drop_prediction/utils/losses.py:64: UserWarning: Using a target size (torch.Size([4, 224, 224])) that is different to the input size (torch.Size([4, 1, 224, 224])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.l1_loss(pred, target, reduction='none')
 25%|██▌       | 25/100 [00:32<01:35,  1.28s/it]/Users/ziyichu/Desktop/Research/CircuitNet/routability_ir_drop_prediction/utils/losses.py:64: UserWarning: Using a target size (torch.Size([1, 224, 224])) that is different to the input size (torch.Size([1, 1, 224, 224])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.l1_loss(pred, target, reduction='none')
 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[26](26/5000): Loss: 5.2908


 26%|██▌       | 26/100 [00:32<01:31,  1.24s/it]


===> Iters[52](52/5000): Loss: 5.2943


 26%|██▌       | 26/100 [00:32<01:31,  1.24s/it]


===> Iters[78](78/5000): Loss: 5.2943


 22%|██▏       | 22/100 [00:27<01:38,  1.26s/it]


===> Iters[100](100/5000): Loss: 4.3980
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_100.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[126](126/5000): Loss: 5.2840


 26%|██▌       | 26/100 [00:33<01:34,  1.27s/it]


===> Iters[152](152/5000): Loss: 5.2736


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[178](178/5000): Loss: 5.2825


 22%|██▏       | 22/100 [00:29<01:45,  1.35s/it]


===> Iters[200](200/5000): Loss: 4.5308
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_200.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[226](226/5000): Loss: 5.2689


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[252](252/5000): Loss: 5.2481


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[278](278/5000): Loss: 5.2502


 22%|██▏       | 22/100 [00:28<01:42,  1.32s/it]


===> Iters[300](300/5000): Loss: 4.4473
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_300.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.27s/it]


===> Iters[326](326/5000): Loss: 5.3036


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[352](352/5000): Loss: 5.2635


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[378](378/5000): Loss: 5.2646


 22%|██▏       | 22/100 [00:29<01:43,  1.32s/it]


===> Iters[400](400/5000): Loss: 4.5030
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_400.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[426](426/5000): Loss: 5.3114


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[452](452/5000): Loss: 5.2435


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[478](478/5000): Loss: 5.3691


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[500](500/5000): Loss: 4.5646
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_500.pth


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[526](526/5000): Loss: 5.2635


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[552](552/5000): Loss: 5.2362


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[578](578/5000): Loss: 5.2556


 22%|██▏       | 22/100 [00:29<01:43,  1.33s/it]


===> Iters[600](600/5000): Loss: 4.5247
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_600.pth


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[626](626/5000): Loss: 5.2689


 26%|██▌       | 26/100 [00:33<01:35,  1.30s/it]


===> Iters[652](652/5000): Loss: 5.2395


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[678](678/5000): Loss: 5.3101


 22%|██▏       | 22/100 [00:29<01:45,  1.36s/it]


===> Iters[700](700/5000): Loss: 4.3920
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_700.pth


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[726](726/5000): Loss: 5.2689


 26%|██▌       | 26/100 [00:34<01:36,  1.31s/it]


===> Iters[752](752/5000): Loss: 5.2502


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[778](778/5000): Loss: 5.2646


 22%|██▏       | 22/100 [00:28<01:42,  1.31s/it]


===> Iters[800](800/5000): Loss: 4.5107
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_800.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[826](826/5000): Loss: 5.2832


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[852](852/5000): Loss: 5.2840


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[878](878/5000): Loss: 5.3117


 22%|██▏       | 22/100 [00:29<01:43,  1.32s/it]


===> Iters[900](900/5000): Loss: 4.5007
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_900.pth


 26%|██▌       | 26/100 [00:37<01:47,  1.45s/it]


===> Iters[926](926/5000): Loss: 5.3274


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[952](952/5000): Loss: 5.2221


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[978](978/5000): Loss: 5.2749


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[1000](1000/5000): Loss: 4.3786
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1000.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.34s/it]


===> Iters[1026](1026/5000): Loss: 5.2342


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[1052](1052/5000): Loss: 5.2747


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[1078](1078/5000): Loss: 5.2832


 22%|██▏       | 22/100 [00:29<01:43,  1.33s/it]


===> Iters[1100](1100/5000): Loss: 4.5274
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1100.pth


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[1126](1126/5000): Loss: 5.2747


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[1152](1152/5000): Loss: 5.3525


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[1178](1178/5000): Loss: 5.3489


 22%|██▏       | 22/100 [00:28<01:42,  1.32s/it]


===> Iters[1200](1200/5000): Loss: 4.4787
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1200.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[1226](1226/5000): Loss: 5.3442


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[1252](1252/5000): Loss: 5.2737


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[1278](1278/5000): Loss: 5.2362


 22%|██▏       | 22/100 [00:29<01:44,  1.33s/it]


===> Iters[1300](1300/5000): Loss: 4.5275
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1300.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[1326](1326/5000): Loss: 5.3183


 26%|██▌       | 26/100 [00:35<01:39,  1.35s/it]


===> Iters[1352](1352/5000): Loss: 5.2730


 26%|██▌       | 26/100 [00:34<01:36,  1.31s/it]


===> Iters[1378](1378/5000): Loss: 5.2816


 22%|██▏       | 22/100 [00:30<01:48,  1.39s/it]


===> Iters[1400](1400/5000): Loss: 4.5115
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1400.pth


 26%|██▌       | 26/100 [00:33<01:36,  1.31s/it]


===> Iters[1426](1426/5000): Loss: 5.2825


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[1452](1452/5000): Loss: 5.2900


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1478](1478/5000): Loss: 5.3340


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[1500](1500/5000): Loss: 4.6364
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1500.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[1526](1526/5000): Loss: 5.3489


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[1552](1552/5000): Loss: 5.2759


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[1578](1578/5000): Loss: 5.2769


 22%|██▏       | 22/100 [00:28<01:40,  1.28s/it]


===> Iters[1600](1600/5000): Loss: 4.3755
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1600.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1626](1626/5000): Loss: 5.2583


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[1652](1652/5000): Loss: 5.2583


 26%|██▌       | 26/100 [00:32<01:32,  1.26s/it]


===> Iters[1678](1678/5000): Loss: 5.2437


 22%|██▏       | 22/100 [00:28<01:41,  1.31s/it]


===> Iters[1700](1700/5000): Loss: 4.5202
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1700.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[1726](1726/5000): Loss: 5.2756


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[1752](1752/5000): Loss: 5.2756


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1778](1778/5000): Loss: 5.2635


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[1800](1800/5000): Loss: 4.3981
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1800.pth


 26%|██▌       | 26/100 [00:32<01:32,  1.26s/it]


===> Iters[1826](1826/5000): Loss: 5.3117


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[1852](1852/5000): Loss: 5.2730


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1878](1878/5000): Loss: 5.2756


 22%|██▏       | 22/100 [00:28<01:39,  1.28s/it]


===> Iters[1900](1900/5000): Loss: 4.4259
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_1900.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1926](1926/5000): Loss: 5.3015


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[1952](1952/5000): Loss: 5.2435


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[1978](1978/5000): Loss: 5.2646


 22%|██▏       | 22/100 [00:28<01:39,  1.28s/it]


===> Iters[2000](2000/5000): Loss: 4.5164
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2000.pth


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[2026](2026/5000): Loss: 5.2749


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2052](2052/5000): Loss: 5.3036


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2078](2078/5000): Loss: 5.2778


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[2100](2100/5000): Loss: 4.4999
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2100.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2126](2126/5000): Loss: 5.2321


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[2152](2152/5000): Loss: 5.2730


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[2178](2178/5000): Loss: 5.2928


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[2200](2200/5000): Loss: 4.5076
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2200.pth


 26%|██▌       | 26/100 [00:32<01:32,  1.26s/it]


===> Iters[2226](2226/5000): Loss: 5.3098


 26%|██▌       | 26/100 [00:32<01:32,  1.25s/it]


===> Iters[2252](2252/5000): Loss: 5.2650


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[2278](2278/5000): Loss: 5.2832


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[2300](2300/5000): Loss: 4.4528
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2300.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2326](2326/5000): Loss: 5.2395


 26%|██▌       | 26/100 [00:33<01:33,  1.27s/it]


===> Iters[2352](2352/5000): Loss: 5.2689


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2378](2378/5000): Loss: 5.2657


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[2400](2400/5000): Loss: 4.4697
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2400.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[2426](2426/5000): Loss: 5.2904


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[2452](2452/5000): Loss: 5.2898


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[2478](2478/5000): Loss: 5.3056


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[2500](2500/5000): Loss: 4.5088
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2500.pth


 26%|██▌       | 26/100 [00:33<01:36,  1.31s/it]


===> Iters[2526](2526/5000): Loss: 5.2778


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[2552](2552/5000): Loss: 5.2898


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[2578](2578/5000): Loss: 5.3999


 22%|██▏       | 22/100 [00:29<01:43,  1.33s/it]


===> Iters[2600](2600/5000): Loss: 4.5206
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2600.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[2626](2626/5000): Loss: 5.3570


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2652](2652/5000): Loss: 5.2435


 26%|██▌       | 26/100 [00:32<01:33,  1.26s/it]


===> Iters[2678](2678/5000): Loss: 5.3697


 22%|██▏       | 22/100 [00:28<01:41,  1.30s/it]


===> Iters[2700](2700/5000): Loss: 4.5333
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2700.pth


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[2726](2726/5000): Loss: 5.3056


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[2752](2752/5000): Loss: 5.2928


 26%|██▌       | 26/100 [00:32<01:33,  1.27s/it]


===> Iters[2778](2778/5000): Loss: 5.2472


 22%|██▏       | 22/100 [00:28<01:41,  1.30s/it]


===> Iters[2800](2800/5000): Loss: 4.5325
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2800.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[2826](2826/5000): Loss: 5.3183


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[2852](2852/5000): Loss: 5.3746


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[2878](2878/5000): Loss: 5.3999


 22%|██▏       | 22/100 [00:28<01:40,  1.29s/it]


===> Iters[2900](2900/5000): Loss: 4.4115
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_2900.pth


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[2926](2926/5000): Loss: 5.3150


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[2952](2952/5000): Loss: 5.2657


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[2978](2978/5000): Loss: 5.3066


 22%|██▏       | 22/100 [00:28<01:41,  1.31s/it]


===> Iters[3000](3000/5000): Loss: 4.4952
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3000.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3026](3026/5000): Loss: 5.3066


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[3052](3052/5000): Loss: 5.2840


 26%|██▌       | 26/100 [00:33<01:33,  1.27s/it]


===> Iters[3078](3078/5000): Loss: 5.3515


 22%|██▏       | 22/100 [00:28<01:41,  1.30s/it]


===> Iters[3100](3100/5000): Loss: 4.4862
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3100.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3126](3126/5000): Loss: 5.2364


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3152](3152/5000): Loss: 5.2620


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3178](3178/5000): Loss: 5.3104


 22%|██▏       | 22/100 [00:28<01:42,  1.32s/it]


===> Iters[3200](3200/5000): Loss: 4.4352
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3200.pth


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[3226](3226/5000): Loss: 5.2609


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3252](3252/5000): Loss: 5.3084


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3278](3278/5000): Loss: 5.3317


 22%|██▏       | 22/100 [00:28<01:41,  1.30s/it]


===> Iters[3300](3300/5000): Loss: 4.5523
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3300.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3326](3326/5000): Loss: 5.3464


 26%|██▌       | 26/100 [00:33<01:34,  1.27s/it]


===> Iters[3352](3352/5000): Loss: 5.3098


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3378](3378/5000): Loss: 5.2749


 22%|██▏       | 22/100 [00:28<01:42,  1.31s/it]


===> Iters[3400](3400/5000): Loss: 4.4212
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3400.pth


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3426](3426/5000): Loss: 5.2943


 26%|██▌       | 26/100 [00:33<01:35,  1.28s/it]


===> Iters[3452](3452/5000): Loss: 5.2657


 26%|██▌       | 26/100 [00:33<01:34,  1.27s/it]


===> Iters[3478](3478/5000): Loss: 5.3705


 22%|██▏       | 22/100 [00:28<01:42,  1.31s/it]


===> Iters[3500](3500/5000): Loss: 4.5771
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3500.pth


 26%|██▌       | 26/100 [00:33<01:36,  1.30s/it]


===> Iters[3526](3526/5000): Loss: 5.4150


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[3552](3552/5000): Loss: 5.2583


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[3578](3578/5000): Loss: 5.2684


 22%|██▏       | 22/100 [00:29<01:43,  1.32s/it]


===> Iters[3600](3600/5000): Loss: 4.4548
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3600.pth


 26%|██▌       | 26/100 [00:33<01:35,  1.29s/it]


===> Iters[3626](3626/5000): Loss: 5.3056


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3652](3652/5000): Loss: 5.2960


 26%|██▌       | 26/100 [00:33<01:34,  1.28s/it]


===> Iters[3678](3678/5000): Loss: 5.3340


 22%|██▏       | 22/100 [00:29<01:43,  1.33s/it]


===> Iters[3700](3700/5000): Loss: 4.5517
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3700.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[3726](3726/5000): Loss: 5.2408


 26%|██▌       | 26/100 [00:33<01:35,  1.30s/it]


===> Iters[3752](3752/5000): Loss: 5.2684


 26%|██▌       | 26/100 [00:35<01:39,  1.35s/it]


===> Iters[3778](3778/5000): Loss: 5.2472


 22%|██▏       | 22/100 [00:31<01:51,  1.43s/it]


===> Iters[3800](3800/5000): Loss: 4.4570
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3800.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[3826](3826/5000): Loss: 5.2960


 26%|██▌       | 26/100 [00:33<01:36,  1.31s/it]


===> Iters[3852](3852/5000): Loss: 5.3442


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[3878](3878/5000): Loss: 5.3705


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[3900](3900/5000): Loss: 4.4694
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_3900.pth


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[3926](3926/5000): Loss: 5.2088


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[3952](3952/5000): Loss: 5.3036


 26%|██▌       | 26/100 [00:34<01:37,  1.31s/it]


===> Iters[3978](3978/5000): Loss: 5.2321


 22%|██▏       | 22/100 [00:29<01:45,  1.36s/it]


===> Iters[4000](4000/5000): Loss: 4.4967
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4000.pth


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4026](4026/5000): Loss: 5.2503


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4052](4052/5000): Loss: 5.3340


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4078](4078/5000): Loss: 5.2883


 22%|██▏       | 22/100 [00:29<01:44,  1.34s/it]


===> Iters[4100](4100/5000): Loss: 4.5436
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4100.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4126](4126/5000): Loss: 5.3746


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4152](4152/5000): Loss: 5.2784


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4178](4178/5000): Loss: 5.3150


 22%|██▏       | 22/100 [00:29<01:46,  1.36s/it]


===> Iters[4200](4200/5000): Loss: 4.4720
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4200.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4226](4226/5000): Loss: 5.2503


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4252](4252/5000): Loss: 5.3036


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4278](4278/5000): Loss: 5.2404


 22%|██▏       | 22/100 [00:30<01:46,  1.37s/it]


===> Iters[4300](4300/5000): Loss: 4.5246
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4300.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4326](4326/5000): Loss: 5.2395


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4352](4352/5000): Loss: 5.2043


 26%|██▌       | 26/100 [00:34<01:37,  1.32s/it]


===> Iters[4378](4378/5000): Loss: 5.2928


 22%|██▏       | 22/100 [00:29<01:45,  1.35s/it]


===> Iters[4400](4400/5000): Loss: 4.5509
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4400.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4426](4426/5000): Loss: 5.3317


 26%|██▌       | 26/100 [00:35<01:39,  1.35s/it]


===> Iters[4452](4452/5000): Loss: 5.3340


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4478](4478/5000): Loss: 5.3515


 22%|██▏       | 22/100 [00:30<01:46,  1.37s/it]


===> Iters[4500](4500/5000): Loss: 4.4621
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4500.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.34s/it]


===> Iters[4526](4526/5000): Loss: 5.3781


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4552](4552/5000): Loss: 5.3664


 26%|██▌       | 26/100 [00:34<01:38,  1.34s/it]


===> Iters[4578](4578/5000): Loss: 5.3525


 22%|██▏       | 22/100 [00:30<01:46,  1.37s/it]


===> Iters[4600](4600/5000): Loss: 4.3906
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4600.pth


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4626](4626/5000): Loss: 5.2650


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4652](4652/5000): Loss: 5.3340


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4678](4678/5000): Loss: 5.3317


 22%|██▏       | 22/100 [00:29<01:45,  1.36s/it]


===> Iters[4700](4700/5000): Loss: 4.5112
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4700.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4726](4726/5000): Loss: 5.3056


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4752](4752/5000): Loss: 5.3697


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4778](4778/5000): Loss: 5.2752


 22%|██▏       | 22/100 [00:29<01:45,  1.36s/it]


===> Iters[4800](4800/5000): Loss: 4.4849
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4800.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.34s/it]


===> Iters[4826](4826/5000): Loss: 5.3664


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4852](4852/5000): Loss: 5.2221


 26%|██▌       | 26/100 [00:35<01:39,  1.35s/it]


===> Iters[4878](4878/5000): Loss: 5.2898


 22%|██▏       | 22/100 [00:29<01:46,  1.36s/it]


===> Iters[4900](4900/5000): Loss: 4.4773
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_4900.pth


 26%|██▌       | 26/100 [00:34<01:38,  1.33s/it]


===> Iters[4926](4926/5000): Loss: 5.2364


 26%|██▌       | 26/100 [00:35<01:39,  1.35s/it]


===> Iters[4952](4952/5000): Loss: 5.2364


 26%|██▌       | 26/100 [00:34<01:39,  1.34s/it]


===> Iters[4978](4978/5000): Loss: 5.2609


 22%|██▏       | 22/100 [00:30<01:46,  1.37s/it]


===> Iters[5000](5000/5000): Loss: 4.5159
Checkpoint saved to ./work_dir/irdrop_mavi//swinTransformer_iters_5000.pth
